LanguageFeatures/Patterns/type_inference_A15_t02.dart

```diff
@@ -0,0 +1,30 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Calculate the static type of the pattern. Using that value type,
+/// recurse through the pattern again downwards to the leaf subpatterns filling
+/// in any missing types in the pattern. This process may also insert implicit
+/// coercions and casts from dynamic when values flow into a pattern during
+/// matching.
+///
+/// @description Check that the calculation of the static type of an identifier
+/// pattern performs casts from dynamic and generic function instantiation
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns
+
+import "../../Utils/expect.dart";
+import "../../Utils/static_type_helper.dart";
+
+T foo<T>(T t) => t;
+
+main() {
+  dynamic pi = 3.14;
+  final (v1 as List<num>) = [pi];
+  v1.expectStaticType<Exactly<List<num>>>();
+
+  final (v2 as int Function(int)) = foo;
+  v2.expectStaticType<Exactly<int Function(int)>>();
+  Expect.equals(foo, v2);
```

Comment:
We don't reach this one, but if we do change the test such that this line is executed then we would presumably not have this equality: As long as `v2` has the type `int Function(int)`, the two functions can't be equal.